### ISSUE TITLE
Closing game while threads are running crashes it

### ISSUE THREAD

**Issue by rainlizard** [godotengine/godot#33054]
**Godot version:** v3.2.alpha3.official

**OS/device including version:** Windows 8.1

**Issue description:**
In 3.2 my game now crashes upon closing it. The issue does not occur in 3.1

When the thread's running and the game is closed it crashes about half the time. You can either call ``get_tree().quit()`` via a keypress outside the thread, or click the X button to close the window to cause the crash.

Calling ``get_tree().quit()`` INSIDE the thread crashes on both 3.1 and 3.2, I think this is probably to be expected since I read something about accessing the scene tree inside threads being a bad idea.

3.2 might now be acting as if ``get_tree().quit()`` is always being called from inside the thread even when executing it outside the thread. Just guessing. I would assume Godot treats closing the game window with the mouse the same as writing that line of code but I dunno.

**Minimal reproduction project:** [Isolate crash upon Escape.zip](https://github.com/godotengine/godot/files/3770264/Isolate.crash.upon.Escape.zip)

**Comment by qarmin** [comment]
Backtrace
```
[1] /lib/x86_64-linux-gnu/libc.so.6(+0x46470) [0x7f9a5146b470] (??:0)
[2] GDScriptFunction::call(GDScriptInstance*, Variant const**, int, Variant::CallError&, GDScriptFunction::CallState*) (/mnt/KubuntuWolne/godot/modules/gdscript/gdscript_function.cpp:1564 (discriminator 1))
[3] GDScriptInstance::call(StringName const&, Variant const**, int, Variant::CallError&) (/mnt/KubuntuWolne/godot/modules/gdscript/gdscript.cpp:1164)
[4] Object::call(StringName const&, Variant const**, int, Variant::CallError&) (/mnt/KubuntuWolne/godot/core/object.cpp:900 (discriminator 1))
[5] _Thread::_start_func(void*) (/mnt/KubuntuWolne/godot/core/bind/core_bind.cpp:2631 (discriminator 1))
[6] ThreadPosix::thread_callback(void*) (/mnt/KubuntuWolne/godot/drivers/unix/thread_posix.cpp:76)
[7] /lib/x86_64-linux-gnu/libpthread.so.0(+0x9669) [0x7f9a5197e669] (??:0)
[8] /lib/x86_64-linux-gnu/libc.so.6(clone+0x43) [0x7f9a51547323] (??:0)
```
Sometimes Godot doesn't crash and show leak
```
WARNING: cleanup: ObjectDB Instances still exist!
   At: core/object.cpp:2069.
Leaked instance: _Thread:1181
Orphan StringName: _thread_function
Orphan StringName: _Thread
StringName: 2 unclaimed string names at exit.
```
or error 
```
ERROR: ~_Thread: Reference to a Thread object object was lost while the thread is still running...
   At: core/bind/core_bind.cpp:2738.
```

**Comment by ludorverr** [comment]
Here's another minimal-reproduction project that will crash reliably.

[Threaded for loop crash.zip](https://github.com/godotengine/godot/files/3803090/Threaded.for.loop.crash.zip)
```
extends Node
var thread = Thread.new()

func _ready():
	thread.start(self, "_thread_function")

func _thread_function(userdata):
	var counter = 0;
	for i in 100000000:
		counter += 1
	print('To produce crash, close the game before this message is printed')
```

It can crash on numbers far lower than 100000000, I'm just using this number for the sake of reproducing the crash every time.

**Comment by timothyqiu** [comment]
FYI, there are many places in `GDScriptFunction::call` that accesses the `GDScriptLanguage` singleton without null check. For example, I always crash at line 1589 here:

https://github.com/godotengine/godot/blob/16fc023d4487c2a3f26b5bdd3827628de2cd0ba2/modules/gdscript/gdscript_function.cpp#L1589-L1596

During `Main::cleanup`, gdscript types are unregistered. The singleton is also deleted and set to `NULL`.

https://github.com/godotengine/godot/blob/16fc023d4487c2a3f26b5bdd3827628de2cd0ba2/modules/gdscript/register_types.cpp#L173-L185

If the singleton is not deleted (commenting out the `if` part above), the reproduction project does not crash. But I don't think it's a right fix.

Should the assumption "Singletons are always available" be true?

**Comment by akien-mga** [comment]
It doesn't seem to crash in the current master branch ff7e7bd260f (on Linux). Instead, it raises these errors:
```
SCRIPT ERROR: : Invalid operands 'int' and 'Nil' in operator '+'.
   At: res://Main.gd:10.
ERROR: ~_Thread: Reference to a Thread object was lost while the thread is still running...
   At: core/bind/core_bind.cpp:2776.
```

**Comment by tom-jk** [comment]
Testing on 3.2.2.stable on linux mint 18.3

- Can reproduce "~_Thread: Reference to a Thread object was lost while the thread is still running..." with "Isolate Crash upon Escape.zip" project. Happens maybe every 1 in 2 runs.
- Can't reproduce crash with the "Threaded for loop crash.zip" project (performed about 20 runs). No crash, errors, leaks in terminal or other apparent issues.

**Comment by qarmin** [comment]
I don't have a crash now but have this errors:
```
SCRIPT ERROR: : Unable to iterate on object of type 'Nil'.
   At: res://Main.gd:21.
ERROR: ~_Thread: Reference to a Thread object was lost while the thread is still running...
   At: core/bind/core_bind.cpp:2835.
```
sometimes leak:
```
Direct leak of 112 byte(s) in 1 object(s) allocated from:
    #0 0x7fe07c401517 in malloc (/lib/x86_64-linux-gnu/libasan.so.6+0xb0517)
    #1 0xec92db3 in Memory::alloc_static(unsigned long, bool) core/os/memory.cpp:82
    #2 0xec92cb2 in operator new(unsigned long, char const*) core/os/memory.cpp:42
    #3 0x50c3433 in ThreadPosix::create_func_posix(void (*)(void*), void*, Thread::Settings const&) drivers/unix/thread_posix.cpp:83
    #4 0xeca98df in Thread::create(void (*)(void*), void*, Thread::Settings const&) core/os/thread.cpp:51
    #5 0xf2b2173 in _Thread::start(Object*, StringName const&, Variant const&, _Thread::Priority) core/bind/core_bind.cpp:2774
    #6 0xf387edb in MethodBind4R<Error, Object*, StringName const&, Variant const&, _Thread::Priority>::call(Object*, Variant const**, int, Variant::CallError&) core/method_bind.gen.inc:3325
    #7 0xe762ee7 in Object::call(StringName const&, Variant const**, int, Variant::CallError&) core/object.cpp:922
    #8 0xe9d9875 in Variant::call_ptr(StringName const&, Variant const**, int, Variant*, Variant::CallError&) core/variant_call.cpp:1127
    #9 0x1a76c4f in GDScriptFunction::call(GDScriptInstance*, Variant const**, int, Variant::CallError&, GDScriptFunction::CallState*) modules/gdscript/gdscript_function.cpp:1088
    #10 0x18aa23f in GDScriptInstance::call(StringName const&, Variant const**, int, Variant::CallError&) modules/gdscript/gdscript.cpp:1237
    #11 0xe762a57 in Object::call(StringName const&, Variant const**, int, Variant::CallError&) core/object.cpp:901
    #12 0xe9d9875 in Variant::call_ptr(StringName const&, Variant const**, int, Variant*, Variant::CallError&) core/variant_call.cpp:1127
    #13 0x1a76c4f in GDScriptFunction::call(GDScriptInstance*, Variant const**, int, Variant::CallError&, GDScriptFunction::CallState*) modules/gdscript/gdscript_function.cpp:1088
    #14 0x18aa23f in GDScriptInstance::call(StringName const&, Variant const**, int, Variant::CallError&) modules/gdscript/gdscript.cpp:1237
    #15 0xe762a57 in Object::call(StringName const&, Variant const**, int, Variant::CallError&) core/object.cpp:901
    #16 0xe7337a4 in MessageQueue::_call_function(Object*, StringName const&, Variant const*, int, bool) core/message_queue.cpp:250
    #17 0xe7344fd in MessageQueue::flush() core/message_queue.cpp:297
    #18 0x962cf94 in SceneTree::iteration(float) scene/main/scene_tree.cpp:483
    #19 0x15c4da4 in Main::iteration() main/main.cpp:2082
    #20 0x14bf6ab in OS_X11::run() platform/x11/os_x11.cpp:3233
    #21 0x143a4e5 in main platform/x11/godot_x11.cpp:56
    #22 0x7fe07af550b2 in __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x270b2)
```
and sometimes invalid read/write
```
==13112==ERROR: AddressSanitizer: heap-use-after-free on address 0x6160000b2f84 at pc 0x000001a832b2 bp 0x7f5ae154a370 sp 0x7f5ae154a360
READ of size 4 at 0x6160000b2f84 thread T6
    #0 0x1a832b1 in GDScriptFunction::call(GDScriptInstance*, Variant const**, int, Variant::CallError&, GDScriptFunction::CallState*) modules/gdscript/gdscript_function.cpp:1470
    #1 0x18aa23f in GDScriptInstance::call(StringName const&, Variant const**, int, Variant::CallError&) modules/gdscript/gdscript.cpp:1237
    #2 0xe762a57 in Object::call(StringName const&, Variant const**, int, Variant::CallError&) core/object.cpp:901
    #3 0xf2b0712 in _Thread::_start_func(void*) core/bind/core_bind.cpp:2728
    #4 0x50c33f3 in ThreadPosix::thread_callback(void*) drivers/unix/thread_posix.cpp:74
    #5 0x7f5afcab0608 in start_thread /build/glibc-YYA7BZ/glibc-2.31/nptl/pthread_create.c:477
    #6 0x7f5afbd17102 in __clone (/lib/x86_64-linux-gnu/libc.so.6+0x122102)

0x6160000b2f84 is located 260 bytes inside of 528-byte region [0x6160000b2e80,0x6160000b3090)
freed by thread T0 here:
    #0 0x7f5afd0c81b7 in __interceptor_free (/lib/x86_64-linux-gnu/libasan.so.6+0xb01b7)
    #1 0xec93ca4 in Memory::free_static(void*, bool) core/os/memory.cpp:178
    #2 0x143b1de in CowData<int>::_unref(void*) core/cowdata.h:213
    #3 0x143af94 in CowData<int>::~CowData() core/cowdata.h:377
    #4 0x143aaf8 in Vector<int>::~Vector() core/vector.h:126
    #5 0x1a91cd3 in GDScriptFunction::~GDScriptFunction() modules/gdscript/gdscript_function.cpp:1796
    #6 0x18f848f in void memdelete<GDScriptFunction>(GDScriptFunction*) core/os/memory.h:117
    #7 0x18a277f in GDScript::~GDScript() modules/gdscript/gdscript.cpp:986
    #8 0x14e80b4 in void memdelete<Reference>(Reference*) core/os/memory.h:117
    #9 0x14da5e4 in Ref<Reference>::unref() core/reference.h:279
    #10 0x14cf514 in Ref<Reference>::~Ref() core/reference.h:295
    #11 0xe826b66 in RefPtr::~RefPtr() core/ref_ptr.cpp:102
    #12 0xe78a419 in Object::~Object() core/object.cpp:1993
    #13 0x958829c in Node::~Node() scene/main/node.cpp:2977
    #14 0xad5ac98 in Spatial::~Spatial() scene/3d/spatial.cpp:847
    #15 0x1e3dd07 in void memdelete<Node>(Node*) core/os/memory.h:117
    #16 0x9519e8f in Node::_notification(int) scene/main/node.cpp:175
    #17 0x175fa99 in Node::_notificationv(int, bool) scene/main/node.h:46
    #18 0x97acc4e in Viewport::_notificationv(int, bool) scene/main/viewport.h:91
    #19 0xe763377 in Object::notification(int, bool) core/object.cpp:932
    #20 0xe752337 in Object::_predelete() core/object.cpp:387
    #21 0xe78a836 in predelete_handler(Object*) core/object.cpp:2053
    #22 0x9687915 in void memdelete<Viewport>(Viewport*) core/os/memory.h:114
    #23 0x9633cad in SceneTree::finish() scene/main/scene_tree.cpp:621
    #24 0x14bfbb8 in OS_X11::run() platform/x11/os_x11.cpp:3237
    #25 0x143a4e5 in main platform/x11/godot_x11.cpp:56
    #26 0x7f5afbc1c0b2 in __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x270b2)

previously allocated by thread T0 here:
    #0 0x7f5afd0c88d0 in __interceptor_realloc (/lib/x86_64-linux-gnu/libasan.so.6+0xb08d0)
    #1 0xec93830 in Memory::realloc_static(void*, unsigned long, bool) core/os/memory.cpp:137
    #2 0x14d8bed in CowData<int>::resize(int) core/cowdata.h:287
    #3 0x14cf114 in Vector<int>::resize(int) core/vector.h:84
    #4 0x14cfc37 in Vector<int>::push_back(int) core/vector.h:152
    #5 0x1984c63 in GDScriptCompiler::_parse_block(GDScriptCompiler::CodeGen&, GDScriptParser::BlockNode const*, int, int, int) modules/gdscript/gdscript_compiler.cpp:1438
    #6 0x1984df6 in GDScriptCompiler::_parse_block(GDScriptCompiler::CodeGen&, GDScriptParser::BlockNode const*, int, int, int) modules/gdscript/gdscript_compiler.cpp:1440
    #7 0x1984df6 in GDScriptCompiler::_parse_block(GDScriptCompiler::CodeGen&, GDScriptParser::BlockNode const*, int, int, int) modules/gdscript/gdscript_compiler.cpp:1440
    #8 0x1988b9f in GDScriptCompiler::_parse_function(GDScript*, GDScriptParser::ClassNode const*, GDScriptParser::FunctionNode const*, bool) modules/gdscript/gdscript_compiler.cpp:1659
    #9 0x19961b1 in GDScriptCompiler::_parse_class_blocks(GDScript*, GDScriptParser::ClassNode const*, bool) modules/gdscript/gdscript_compiler.cpp:2023
    #10 0x199a4a0 in GDScriptCompiler::compile(GDScriptParser const*, GDScript*, bool) modules/gdscript/gdscript_compiler.cpp:2175
    #11 0x1893f3a in GDScript::reload(bool) modules/gdscript/gdscript.cpp:605
    #12 0x18d10e7 in ResourceFormatLoaderGDScript::load(String const&, String const&, Error*) modules/gdscript/gdscript.cpp:2293
    #13 0xf148691 in ResourceLoader::_load(String const&, String const&, String const&, bool, Error*) core/io/resource_loader.cpp:270
    #14 0xf14b313 in ResourceLoader::load(String const&, String const&, bool, Error*) core/io/resource_loader.cpp:401
    #15 0xc384e61 in ResourceInteractiveLoaderText::poll() scene/resources/resource_format_text.cpp:433
    #16 0xf1429aa in ResourceFormatLoader::load(String const&, String const&, Error*) core/io/resource_loader.cpp:197
    #17 0xf148691 in ResourceLoader::_load(String const&, String const&, String const&, bool, Error*) core/io/resource_loader.cpp:270
    #18 0xf14b313 in ResourceLoader::load(String const&, String const&, bool, Error*) core/io/resource_loader.cpp:401
    #19 0x15beba5 in Main::start() main/main.cpp:1938
    #20 0x143a418 in main platform/x11/godot_x11.cpp:55
    #21 0x7f5afbc1c0b2 in __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x270b2)

Thread T6 created by T0 here:
    #0 0x7f5afd06fa95 in __interceptor_pthread_create (/lib/x86_64-linux-gnu/libasan.so.6+0x57a95)
    #1 0x50c38e8 in ThreadPosix::create_func_posix(void (*)(void*), void*, Thread::Settings const&) drivers/unix/thread_posix.cpp:90
    #2 0xeca98df in Thread::create(void (*)(void*), void*, Thread::Settings const&) core/os/thread.cpp:51
    #3 0xf2b2173 in _Thread::start(Object*, StringName const&, Variant const&, _Thread::Priority) core/bind/core_bind.cpp:2774
    #4 0xf387edb in MethodBind4R<Error, Object*, StringName const&, Variant const&, _Thread::Priority>::call(Object*, Variant const**, int, Variant::CallError&) core/method_bind.gen.inc:3325
    #5 0xe762ee7 in Object::call(StringName const&, Variant const**, int, Variant::CallError&) core/object.cpp:922
    #6 0xe9d9875 in Variant::call_ptr(StringName const&, Variant const**, int, Variant*, Variant::CallError&) core/variant_call.cpp:1127
    #7 0x1a76c4f in GDScriptFunction::call(GDScriptInstance*, Variant const**, int, Variant::CallError&, GDScriptFunction::CallState*) modules/gdscript/gdscript_function.cpp:1088
    #8 0x18aa23f in GDScriptInstance::call(StringName const&, Variant const**, int, Variant::CallError&) modules/gdscript/gdscript.cpp:1237
    #9 0xe762a57 in Object::call(StringName const&, Variant const**, int, Variant::CallError&) core/object.cpp:901
    #10 0xe9d9875 in Variant::call_ptr(StringName const&, Variant const**, int, Variant*, Variant::CallError&) core/variant_call.cpp:1127
    #11 0x1a76c4f in GDScriptFunction::call(GDScriptInstance*, Variant const**, int, Variant::CallError&, GDScriptFunction::CallState*) modules/gdscript/gdscript_function.cpp:1088
    #12 0x18aa23f in GDScriptInstance::call(StringName const&, Variant const**, int, Variant::CallError&) modules/gdscript/gdscript.cpp:1237
    #13 0xe762a57 in Object::call(StringName const&, Variant const**, int, Variant::CallError&) core/object.cpp:901
    #14 0xe7337a4 in MessageQueue::_call_function(Object*, StringName const&, Variant const*, int, bool) core/message_queue.cpp:250
    #15 0xe7344fd in MessageQueue::flush() core/message_queue.cpp:297
    #16 0x962cf94 in SceneTree::iteration(float) scene/main/scene_tree.cpp:483
    #17 0x15c4da4 in Main::iteration() main/main.cpp:2082
    #18 0x14bf6ab in OS_X11::run() platform/x11/os_x11.cpp:3233
    #19 0x143a4e5 in main platform/x11/godot_x11.cpp:56
    #20 0x7f5afbc1c0b2 in __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x270b2)
```

**Comment by rainlizard** [comment]
To make matters even more confusing I tested on my other machine and I got some different results.

- v3.2.2-stable_win64 Windows 8.1 64bit 4790k CPU

"Threaded for loop crash" still crashing 100% of the time and the first project "Isolate crash upon Escape" crashes about 90% of the time.

- v3.2.2-stable_win64 Windows 10 64bit e8400 CPU (11 year old PC)
No crashing at all. I was surprised, I thought this was a Windows vs Linux issue.

To be clear, here's what the "crash" looks like, upon closing the game.
![crash1](https://user-images.githubusercontent.com/15337628/87777498-f1625d00-c86c-11ea-93a3-99dd99c6f132.png)
![crash2](https://user-images.githubusercontent.com/15337628/87777505-f32c2080-c86c-11ea-9437-a2cc5bf1ac97.png)
